Create nullable migration tutorial (#10383)
dotnet/docs: docs/csharp/nullable-references.md

- **A reference can be null**. No warnings are issued when a reference type is initialized to null, or later assigned to null.
- **A reference is assumed to be not null**. The compiler doesn't issue any warnings when reference types are dereferenced. (With nullable references,  the compiler issues warnings whenever you dereference a variable that may be null).

> With the addition of nullable reference types, you can declare your intent more clearly. The `null` value is the correct way to represent that a variable does not refer to a value. Don't use this feature to remove all `null` values from your code. Rather, you should declare your intent to the compiler and other developers that read your code. By declaring your intent, the compiler informs you when you write code that is inconsistent with that intent.

Reviewer: ```suggestion
With the addition of nullable reference types, you can declare your intent more clearly. The `null` value is the correct way to represent that a variable doesn't refer to a value. Don't use this feature to remove all `null` values from your code. Rather, you should declare your intent to the compiler and other developers that read your code. By declaring your intent, the compiler informs you when you write code that is inconsistent with that intent.
```

You can read details about this operator in the [draft nullable reference types](https://github.com/dotnet/csharplang/blob/master/proposals/nullable-reference-types-specification.md#the-null-forgiving-operator) specification proposal on GitHub.

## Nullability of types

Any reference type can have one of four *nullabilities*, which describes when warnings are generated:

> - *Nonnullable*: Null cannot be assigned to variables of this type. Variables of this type do not need to be null-checked before dereferencing.

Reviewer: ```suggestion
- *Nonnullable*: Null can't be assigned to variables of this type. Variables of this type don't need to be null-checked before dereferencing.
```

- *Nullable*: Null can be assigned to variables of this type. Dereferencing variables of this type without first checking for `null` causes a warning.
- *Oblivious*: This is the pre-C# 8 state. Variables of this type can be dereferenced or assigned without warnings.

> - *Unknown*: This is generally for type parameters where constraints do not tell the compiler that the type must be *nullable* or *nonnullable*.

Reviewer: ```suggestion
- *Unknown*: This is generally for type parameters where constraints don't tell the compiler that the type must be *nullable* or *nonnullable*.
```

- `#nullable safeonly`: Set the nullable annotation context to **enabled**, and the warning context to **safeonly**.
- `#nullable restore`: Restores the nullable annotation context and nullable warning context to the project settings.
- `#pragma warning disable nullable`: Set the nullable warning context to **disabled**.
- `#pragma warning enable nullable`: Set the nullable warning context to **enabled**.
- `#pragma warning restore nullable`: Restores the nullable warning context to the project settings.
- `#pragma warning safeonly nullable`: Sets the nullable warning context to **safeonly**.

The default nullable annotation and warning contexts are `disabled`. That decision means that your existing code compiles without changes and without generating any new warnings.

The differences between the `enabled` and `safeonly` nullable warning contexts are warnings for assigning a nullable reference to a non-nullable reference. The following assignment generates a warning in an `enabled` warning context, but not a `safeonly` warning context. However, the second line, where `s` is dereferenced, generates a warning in a `safeonly` context:

```csharp

> string s = null; // warning when context is enabled.

Reviewer 2: It might be worth explicitly saying it's the *warning* context here and in the next line, given that there are two different contexts that could be referred to.

The behavior is the same as previous versions of C#.

The compiler uses the following rules in an enabled nullable annotation context:

- Any variable of a reference type is a **non-nullable reference**.
- Any non-nullable reference may be dereferenced safely.
- Any nullable reference type (noted by `?` after the type in the variable declaration) may be null. Static analysis determines if the value is known to be non-null when it is dereferenced. If not, the compiler warns you.

> - You can use the null-forgiving operator to declare that a nullable reference isn't null.

Reviewer 2: "null-forgiving" or "null forgiveness"? Both are used quite a bit here.

Author: The spec is standardizing on "null forgiving". I did the same.

> In an enabled nullable annotation context, the `?` character appended to a reference type declares a **nullable reference type**. The **null forgiveness operator** (`!`) may be appended to an expression to declare that the expression is not null.

Reviewer 2: Does the `!` operator have an expression type which is always non-null as well, or does it only affect the static analysis?

Author: When I read the [spec](https://github.com/dotnet/csharplang/blob/master/proposals/nullable-reference-types-specification.md#the-null-forgiving-operator) I think it does not change the expression type from `nullable_reference_type` to `nonnullable_reference_type`. It only declares that the variable is currently "not null".

  /cc @MadsTorgersen for verification.

> In an enabled nullable annotation context, the `?` character appended to a reference type declares a **nullable reference type**. The **null forgiveness operator** (`!`) may be appended to an expression to declare that the expression is not null.

Reviewer: ```suggestion
In an enabled nullable annotation context, the `?` character appended to a reference type declares a **nullable reference type**. The **null forgiveness operator** (`!`) may be appended to an expression to declare that the expression isn't null.
```

## Nullable warning context

> The nullable warning context is distinct from the nullable annotation context. Warnings can be enabled even when the new annotations are disabled. The compiler uses static flow analysis to determine the **null state** of any reference. The null state is either **not null** or **maybe null** when the *nullable warning context* is not **disabled**. If you dereference a reference when the compiler has determined it's **maybe null**, the compiler warns you. The state of a reference is **maybe null** unless the compiler can determine one of two conditions:

Reviewer: ```suggestion
The nullable warning context is distinct from the nullable annotation context. Warnings can be enabled even when the new annotations are disabled. The compiler uses static flow analysis to determine the **null state** of any reference. The null state is either **not null** or **maybe null** when the *nullable warning context* isn't **disabled**. If you dereference a reference when the compiler has determined it's **maybe null**, the compiler warns you. The state of a reference is **maybe null** unless the compiler can determine one of two conditions:
```

1. The variable has been definitely assigned to a non-null value.

> 1. The variable has been checked against null before de-referencing it.

Reviewer 2: It may be worth mentioning expressions instead of variables. The fact that `if (foo.Address != null)` changes whether I can use `foo.Address.Country` later without a warning was one of the aspects of nullability that most surprised me the first time I saw it.
